File: go/pe023.go
package main

import (
	"fmt"
	"math"
)

func sum_factors(num int) int {
	asum := 1
	for i := 2; i <= int(math.Sqrt(float64(num))); i++ {
		if num%i == 0 {
			divisor := num / i
			asum += i
			if i != divisor {
				asum += divisor
			}
		}
	}
	return asum
}

func main() {
	const MAX = 28123
	abundant := []int{}
	amap := make(map[int]int)
	for i := 1; i < MAX; i++ {
		fsum := sum_factors(i)
		if i < fsum {
			abundant = append(abundant, i)
			amap[i] = fsum
		}
	}

	var asum int
	for i := 0; i < abundant[0]; i++ {
		asum += i
	}
	var addi bool
	for i := abundant[0]; i < MAX; i++ {
		addi = true
		for _, j := range abundant {
			if j < i {
				diff := i - j
				if _, ok := amap[diff]; ok {
					addi = false
					break
				}
			}
		}
		if addi {
			asum += i
		}
	}
	fmt.Printf("%d\n", asum)
}

File: ruby/pe023.rb
#!/usr/bin/ruby

def sum_factors(number)
  asum = 1
  limit = Math.sqrt(number).to_i
  (2 .. limit).each do |i|
    if number % i == 0
      divisor = number / i
      asum += i
      if i != divisor
        asum += divisor
      end
    end
  end
  return asum
end

MAX = 28123
abundant = Array.new
adictionary = Hash.new

(1 ... MAX).each do |i|
  fsum = sum_factors(i)
  if i < fsum
    abundant.push(i)
    adictionary[i] = fsum
  end
end

asum = 0
(1 ... abundant[0]).each do |i|
  asum += i
end
(abundant[0] ... MAX).each do |i|
  addi = true
  abundant.each do |j|
    if j >= i
      break
    end
    diff = i - j
    if adictionary.has_key?(diff)
      addi = false
      break
    end
  end
  if addi
    asum += i
  end
end

puts(asum)
